src/Data/TypeRep/CacheMap.hs

```diff
@@ -0,0 +1,147 @@
+{-# LANGUAGE BangPatterns   #-}
+{-# LANGUAGE GADTs          #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE MagicHash      #-}
+{-# LANGUAGE PolyKinds      #-}
+{-# LANGUAGE TypeFamilies   #-}
+
+module Data.TypeRep.CacheMap
+       ( -- * Map type
+         TypeRepMap (..)
+
+         -- 'TypeRepMap' interface
+       , empty
+       , insert
+       , lookup
+       , size
+
+         -- * Helpful testing functions
+       , TF (..)
+       , fromList
+       ) where
+
+import Prelude hiding (lookup)
+
+import Control.Arrow ((&&&))
+import Data.Kind (Type)
+import Data.Proxy (Proxy (..))
+import Data.Typeable (Typeable, typeRep, typeRepFingerprint)
+import Data.Word (Word64)
+import GHC.Base (Any, Int (..), Int#, uncheckedIShiftRA#, (*#), (+#), (-#), (<#), (==#))
+import GHC.Exts (inline, sortWith)
+import GHC.Fingerprint (Fingerprint (..))
+import GHC.Prim
+import GHC.Word (Word64 (..))
+import Unsafe.Coerce (unsafeCoerce)
+
+import qualified Data.Vector as V
+import qualified Data.Vector.Unboxed as Unboxed
+
+data TypeRepMap (f :: k -> Type) = TypeRepMap
+    { fingerprintAs :: Unboxed.Vector Word64
+    , fingerprintBs :: Unboxed.Vector Word64
+    , anys          :: V.Vector Any
+    }
+
+fromAny :: Any -> f a
+fromAny = unsafeCoerce
+
+-- | Empty structure.
+empty :: TypeRepMap f
+empty = TypeRepMap mempty mempty mempty
+
+-- | Inserts the value with its type as a key.
+insert :: forall a f . Typeable a => a -> TypeRepMap f -> TypeRepMap f
+insert = undefined
+
+-- | Looks up the value at the type.
+-- >>> let x = lookup $ insert (11 :: Int) empty
+-- >>> x :: Maybe Int
+-- Just 11
+-- >>> x :: Maybe ()
+-- Nothing
+lookup :: forall a f . Typeable a => TypeRepMap f -> Maybe (f a)
+lookup tVect =  fromAny . (anys tVect V.!)
+            <$> cachedBinarySearch (typeRepFingerprint $ typeRep $ Proxy @a)
+                             (fingerprintAs tVect)
+                             (fingerprintBs tVect)
+
+-- | Returns the size of the 'TypeRepMap'.
+size :: TypeRepMap f -> Int
+size = Unboxed.length . fingerprintAs
+
+-- | Binary searched based on this article
+-- http://bannalia.blogspot.com/2015/06/cache-friendly-binary-search.html
+-- with modification for our two-vector search case.
+cachedBinarySearch :: Fingerprint -> Unboxed.Vector Word64 -> Unboxed.Vector Word64 -> Maybe Int
+cachedBinarySearch (Fingerprint (W64# a) (W64# b)) fpAs fpBs = inline (go 0#)
+  where
+    go :: Int# -> Maybe Int
+    go i = case i <# len of
+        0# -> Nothing
+        _  -> let !(W64# valA) = Unboxed.unsafeIndex fpAs (I# i) in case a `ltWord#` valA of
+            0#  -> case a `eqWord#` valA of
+                0# -> go (2# *# i +# 2#)
+                _ -> let !(W64# valB) = Unboxed.unsafeIndex fpBs (I# i) in case b `ltWord#` valB of
```

Comment:
I think this part can be optimized a bit. In almost all cases (as far as I understand) if first halves of of fingreprint are equal, second halves will be equal as well. So we don't need to perform extra redundant `ltWord#` in almost all cases. And thus we can check for `eqWord#` first.

Reply:
Thanks!
It's improved for the tiny amount (~ 6 ns)